cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/cl_kernels/df_batch_predict_classification_kernels.cl

```diff
@@ -0,0 +1,197 @@
+/* file: df_batch_classification_kernels.cl */
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Implementation of decision forest Batch classification OpenCL kernels.
+//--
+*/
+
+#ifndef __DF_BATCH_PREDICT_CLASSIFICATION_KERNELS_CL__
+#define __DF_BATCH_PREDICT_CLASSIFICATION_KERNELS_CL__
+
+#include <string.h>
+
+#define DECLARE_SOURCE(name, src) static const char * name = #src;
+
+DECLARE_SOURCE(
+    df_batch_predict_classification_kernels,
+    __kernel void predictByTreesWeighted(const __global algorithmFPType * data, const __global int * ftrIdx,
+                                         const __global int * classLabelsOrNextNodeIdx, const __global algorithmFPType * ftrValue,
+                                         const __global double * classProba, __global algorithmFPType * obsClassHist, algorithmFPType scale,
+                                         int nRows, int nCols, int nTrees, uint maxTreeSize, int treeOffset) {
+        const int nClasses      = NUM_OF_CLASSES;
+        const int local_id      = get_local_id(0);
+        const int local_size    = get_local_size(0);
+        const int n_groups      = get_num_groups(0);
+        const int group_id      = get_group_id(0);
+        const int n_tree_groups = get_num_groups(1);
+        const int tree_group_id = get_group_id(1);
+        const int tree_id       = treeOffset + tree_group_id;
+
+        const int nElementsForGroup = nRows / n_groups + !!(nRows % n_groups);
+
+        int iStart = group_id * nElementsForGroup;
+        int iEnd   = (group_id + 1) * nElementsForGroup;
+
+        iEnd = (iEnd > nRows) ? nRows : iEnd;
+
+        if (tree_id < nTrees)
+        {
+            const __global int * ftrIdxForTree                   = ftrIdx + tree_id * maxTreeSize;
+            const __global int * classLabelsOrNextNodeIdxForTree = classLabelsOrNextNodeIdx + tree_id * maxTreeSize;
+            const __global algorithmFPType * ftrValueForTree     = ftrValue + tree_id * maxTreeSize;
+            const __global double * classProbaForTree            = classProba + tree_id * maxTreeSize * nClasses;
+
+            uint treeRootIsSplit = (uint)(-1 != ftrIdxForTree[0]);
```

Comment:
what `ftr` stands for? `-1` - magic number

Reply:
leaf node, fixed